lib/rrd.ml

```diff
@@ -692,55 +692,76 @@ let xml_to_output rrd output =
       do_rras rrd.rrd_rras output)
     output
 
-
-let json_to_string rrd =
-  let do_dss ds_list =
-    "ds:["^(String.concat "," (List.map (fun ds ->
-        "{name:\""^ds.ds_name^"\",type:\""^(match ds.ds_ty with Gauge -> "GAUGE" | Absolute -> "ABSOLUTE" | Derive -> "DERIVE")^
-        "\",minimal_heartbeat:" ^(Utils.f_to_s ds.ds_mrhb)^",min:"^(Utils.f_to_s ds.ds_min)^
-        ",max:"^(Utils.f_to_s ds.ds_max)^",last_ds:0.0,value:0.0,unknown_sec:0}") ds_list))^"]"
-  in
-
-  let do_rra_cdps cdp_list =
-    "ds:["^(String.concat "," (List.map (fun cdp ->
-        "{primary_value:0.0,secondary_value:0.0,value:"^(Utils.f_to_s cdp.cdp_value)^
-        ",unknown_datapoints:"^(Printf.sprintf "%d" cdp.cdp_unknown_pdps)^"}") cdp_list))^"]"
-  in
-
-  let do_database rings =
-    if Array.length rings = 0 then "[]" else
+module Json = struct
+
+  let string fmt = Printf.ksprintf (fun msg -> `String msg) fmt
+  let float x    = string "%.2f" x
+  let record xs  = `O xs
+  let array  xs  = `A xs
+
+  let ty = function
+    | Gauge     -> string "GAUGE"
+    | Absolute  -> string "ABSOLUTE"
+    | Derive    -> string "DERIVE"
+
+  let datasource ds =
+    record
+      [ "name"               , string "%s" ds.ds_name
+      ; "type"               , ty ds.ds_ty
+      ; "minimal_hearbeat"   , string "%s" (Utils.f_to_s ds.ds_mrhb)
+      ; "min"                , string "%s" (Utils.f_to_s ds.ds_min)
+      ; "max"                , string "%s" (Utils.f_to_s ds.ds_max)
+      ; "last_ds",           ( match ds.ds_last with
+                             | VT_Float x -> string "%s" (Utils.f_to_s x)
```

Comment:
Here is an example RRD json: https://github.com/oetiker/rrdtool-1.x/blob/f4040b2c1b9e51fe23b8bc546b43e0b33440f855/tests/xport1.json.output

Reply:
We can defer this to a future PR